src/LibraryManager.Vsix/Shared/LibraryCommandService.cs

```diff
@@ -0,0 +1,352 @@
+            }
+        }
+
+        private string GetTaskTitle(OperationType operation, string libraryId)
+        {
+            switch (operation)
+            {
+                case OperationType.Restore:
+                    {
+                        return LibraryManager.Resources.Text.Restore_OperationStarted;
+                    }
+                case OperationType.Clean:
+                    {
+                        return LibraryManager.Resources.Text.Clean_OperationStarted;
+                    }
+                case OperationType.Uninstall:
+                    {
+                        return string.Format(LibraryManager.Resources.Text.Uninstall_LibraryStarted, libraryId ?? string.Empty);
+                    }
+                case OperationType.Upgrade:
+                    {
+                        return string.Format(LibraryManager.Resources.Text.Update_LibraryStarted, libraryId ?? string.Empty);
+                    }
+            }
+
+            return string.Empty;
+        }
+
+        private void PostRestoreTelemetryData(IEnumerable<ILibraryOperationResult> results, TimeSpan elapsedTime)
+        {
+            var telResult = new Dictionary<string, double>();
+            foreach (ILibraryOperationResult result in results.Where(r => r.Success))
+            {
+                if (result.InstallationState.ProviderId != null)
+                {
+                    telResult.TryGetValue(result.InstallationState.ProviderId, out double count);
+                    telResult[result.InstallationState.ProviderId] = count + 1;
+                }
+            }
+            telResult.Add("time", elapsedTime.TotalMilliseconds);
+            Telemetry.TrackUserTask("restore", TelemetryResult.None, telResult.Select(i => new KeyValuePair<string, object>(i.Key, i.Value)).ToArray());
+        }
+
+        private void AddErrorsToErrorList(string projectName, string configFile, IEnumerable<ILibraryOperationResult> results)
+        {
+            var errorList = new ErrorList(projectName, configFile);
+            errorList.HandleErrors(results);
+        }
+
+        private async Task AddFilesToProjectAsync(string configFilePath, Project project, IEnumerable<ILibraryOperationResult> results, CancellationToken cancellationToken)
+        {
+            string cwd = Path.GetDirectoryName(configFilePath);
+            var files = new List<string>();
+
+            foreach (ILibraryOperationResult state in results)
+            {
+                if (state.Success)
+                {
+                    IEnumerable<string> absoluteFiles = state.InstallationState.Files
+                        .Select(file => Path.Combine(cwd, state.InstallationState.DestinationPath, file)
+                        .Replace('/', Path.DirectorySeparatorChar));
+                    files.AddRange(absoluteFiles.Where(file => !files.Contains(file)));
```

Comment:
How big are you expecting this list to get? If this can get fairly large, the Contains call could get expensive.

Reply:
I did not originally write this line of code, but this list can get really big. We have code now that detects duplicates before the actual install so we should not end up on a situation where we have duplicates. I think I can replace it by files.AddRange(absoluteFiles);

Reply:
Also this code should only run if project != null